:focus-within pseudo class doesn't get invalidated when frame loses focus

rniwa opened the pull request:
#### ac8282feee7f212d45261661d82051c83d344cc2
<pre>
:focus-within pseudo class doesn&apos;t get invalidated when frame loses focus
<a href="https://bugs.webkit.org/show_bug.cgi?id=242526">https://bugs.webkit.org/show_bug.cgi?id=242526</a>

Reviewed by Antti Koivisto and Darin Adler.

This patch fixes the bug that :focus-within pseudo fails to start applying or stop applying upon frame
gaining or losing focus, and makes pseudo class invalidation more precise.

A new boolean flag FrameSelection::m_isActive, which mirrors the value of FocusController::isActive()
while the frame associated with FrameSelection is focused, is introduced and used in FrameSelection&apos;s
isFocusedAndActive on behalf of FocusController::isActive(). It&apos;s used to compute the value of :focus,
:focus-visible, :focus-within and pseudo classes. Most notably, FrameSelection::pageActivationChanged
and FrameSelection::setFocused use it to flip the flag with Style::PseudoClassChangeInvalidation.

Note that the value of FrameSelection::m_isActive could be stale when the frame associated with
FrameSelection is no longer focused. This isn&apos;t an issue in practice since it means isFocusedAndActive
would always return false in such cases (since m_focused check fails).

* LayoutTests/fast/shadow-dom/focus-ring-on-shadow-host-focuswithin-expected.html: Added.
* LayoutTests/fast/shadow-dom/focus-ring-on-shadow-host-focuswithin.html: Added.
* LayoutTests/fast/shadow-dom/focus-ring-on-shadow-host-sibling-expected.html: Added.
* LayoutTests/fast/shadow-dom/focus-ring-on-shadow-host-sibling.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const): Fixed a bug that this was checking hasFocusWithin
without isFrameFocused.
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::isPageActive): Added.
(WebCore::FrameSelection::FrameSelection): Initialize newly introduced m_isActive.
(WebCore::FrameSelection::focusedOrActiveStateChanged): Moved the logic to invalidate elements to
callers of this function: FrameSelection::pageActivationChanged and FrameSelection::setFocused.
(WebCore::invalidateFocusedElementAndShadowIncludingAncestors): Added.
(WebCore::FrameSelection::pageActivationChanged):
(WebCore::FrameSelection::setFocused):
(WebCore::FrameSelection::isFocusedAndActive const):
* Source/WebCore/editing/FrameSelection.h:

Canonical link: <a href="https://commits.webkit.org/252324@main">https://commits.webkit.org/252324@main</a>
</pre>

darinadler commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -166,6 +167,7 @@ FrameSelection::FrameSelection(Document* document)
     , m_caretPaint(true)
     , m_isCaretBlinkingSuspended(false)
     , m_focused(document && document->frame() && document->page() && document->page()->focusController().focusedFrame() == document->frame())
+    , m_isActive(m_document && m_document->page() && m_document->page()->focusController().isActive())
```
Should we make a function that’s shared for this expression? I know this is not a lot of code, but I think it would be nice to have a named function to emphasize that it’s the same all three times.

rniwa replied:
Added `isPageActive`.

darinadler reviewed (approved):
I think this is excellent and correct now; of course, @anttijk should still look it over, since he’s the one who spotted it was wrong last time.

anttijk reviewed (approved):
Looks good!

rniwa commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2142,35 +2144,51 @@ void FrameSelection::focusedOrActiveStateChanged()
+        bool isActive = m_document->page() && m_document->page()->focusController().isActive();
+        auto invalidations = invalidateFocusedElementAndShadowIncludingAncestors(*focusedElement, m_focused && isActive);
+        m_isActive = isActive;
+    }
```
Oops, I need to update the flags even when focused element is nullptr here.

rniwa commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2142,35 +2144,51 @@ void FrameSelection::focusedOrActiveStateChanged()
+        auto invalidations = invalidateFocusedElementAndShadowIncludingAncestors(*focusedElement, isFocused && isActive);
+        m_focused = isFocused;
+        m_isActive = isActive;
+    }
```
Ditto.

rniwa committed:
:focus-within pseudo class doesn't get invalidated when frame loses focus
https://bugs.webkit.org/show_bug.cgi?id=242526

Reviewed by Antti Koivisto and Darin Adler.

This patch fixes the bug that :focus-within pseudo fails to start applying or stop applying upon frame
gaining or losing focus, and makes pseudo class invalidation more precise.

A new boolean flag FrameSelection::m_isActive, which mirrors the value of FocusController::isActive()
while the frame associated with FrameSelection is focused, is introduced and used in FrameSelection's
isFocusedAndActive on behalf of FocusController::isActive(). It's used to compute the value of :focus,
:focus-visible, :focus-within and pseudo classes. Most notably, FrameSelection::pageActivationChanged
and FrameSelection::setFocused use it to flip the flag with Style::PseudoClassChangeInvalidation.

Note that the value of FrameSelection::m_isActive could be stale when the frame associated with
FrameSelection is no longer focused. This isn't an issue in practice since it means isFocusedAndActive
would always return false in such cases (since m_focused check fails).

* LayoutTests/fast/shadow-dom/focus-ring-on-shadow-host-focuswithin-expected.html: Added.
* LayoutTests/fast/shadow-dom/focus-ring-on-shadow-host-focuswithin.html: Added.
* LayoutTests/fast/shadow-dom/focus-ring-on-shadow-host-sibling-expected.html: Added.
* LayoutTests/fast/shadow-dom/focus-ring-on-shadow-host-sibling.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const): Fixed a bug that this was checking hasFocusWithin
without isFrameFocused.
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::isPageActive): Added.
(WebCore::FrameSelection::FrameSelection): Initialize newly introduced m_isActive.
(WebCore::FrameSelection::focusedOrActiveStateChanged): Moved the logic to invalidate elements to
callers of this function: FrameSelection::pageActivationChanged and FrameSelection::setFocused.
(WebCore::invalidateFocusedElementAndShadowIncludingAncestors): Added.
(WebCore::FrameSelection::pageActivationChanged):
(WebCore::FrameSelection::setFocused):
(WebCore::FrameSelection::isFocusedAndActive const):
* Source/WebCore/editing/FrameSelection.h:

Canonical link: https://commits.webkit.org/252324@main
